base/src/main/java/io/spine/code/proto/ref/Parsing.java

```diff
@@ -0,0 +1,61 @@
+/*
+ * Copyright 2019, TeamDev. All rights reserved.
+ *
+ * Redistribution and use in source and/or binary forms, with or without
+ * modification, must retain the above copyright notice and the following
+ * disclaimer.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+package io.spine.code.proto.ref;
+
+import com.google.common.collect.ImmutableList;
+
+import java.util.Optional;
+import java.util.function.Function;
+
+/**
+ * Attempts to parse the passed value by sequentially invoking provider functions.
+ */
+final class Parsing {
+
+    private final String value;
+    private final ImmutableList<Provider> provider;
+
+    Parsing(String value, Provider... provider) {
+        this.value = value;
+        this.provider = ImmutableList.copyOf(provider);
+    }
+
+    Optional<TypeRef> parse() {
+        for (Provider supplier : provider) {
+            Optional<TypeRef> found = supplier.apply(value);
+            if (found.isPresent()) {
+                return found;
+            }
+        }
+        return Optional.empty();
+    }
+
+    /**
+     * Provides a type reference <em>if</em> the passed string matches the format
+     * requirements of the provider, otherwise returns an empty {@code Optional}.
+     *
+     * @see CompositeTypeRef#parsePart(String)
+     * @see TypeRef#parse(String)
+     */
+    @FunctionalInterface
+    interface Provider extends Function<String, Optional<TypeRef>> {
```

Comment:
Maybe it should be `Parser` or `TypeRefParser`?  Also, is there any particular reason to extend from `Function`? Having a `@FunctionalInterface` itself should be sufficient and we can provide a cleaner method than a general-purpose `apply`.
The following looks cleaner to me:
```java
@FunctionalInterface
interface TypeRefParser {
     Optional<TypeRef> parse(String reference);
}
```

Reply:
Well spotted!  This way it is simpler.